### PR TITLE
[Persistence] Consolidate common behaviour between `Pool` and `Account` into a shared interface (Issue #102)

### LINKED ISSUE #102: [Persistence] Consolidate common behaviour between `Pool` and `Account` into a shared interface
# Objective
Remove redundant code used by the `Pool` and `Account` actors in the persistence module introduced in https://github.com/pokt-network/pocket/pull/73.

This is a good starter task opportunity to get acquainted with the codebase in the persistence module.

# Origin Document

From the [utility specification[https://github.com/pokt-network/pocket-network-protocol/tree/main/utility), a `Pool` is described as:
```
	A ModulePool is a particular type that though similar in structure to an Account, the
	functionality of each is quite specialized to its use case. These pools are maintained by
	the protocol and are completely autonomous, owned by no actor on the network. Unlike Accounts,
	tokens are able to be directly minted to and burned from ModulePools. Examples of ModuleAccounts
	include StakingPools and the FeeCollector
```			

Similar to how the functionality of Fisherman, ServiceNode, Validator and Application was consolidated via `persistence/schema/protocol_actor.go` and `persistence/schema/base_actor.go`, the functionality of Pool and Account can also be shared to remove redundant code.

# Goals
## Deliverables
- [ ] Extract common functionality of `Pool` and `Account` into a common interface
- [ ] At a minimum, the following files will need to be affected: `persistence/schema/account.go`, `persistence/account.go`, `persistence/test/account_test.go`

## Testing Methodology
- [ ] ` $ make test_persistence`
- [ ] ` $ make test_all`
- [ ] Run a `LocalNet` following the instructions in `docs/development/README.md`

## Non-goals
- Adding new functionality to the persistence module

Creator: @Olshansk 
Owner: @DragonDmoney
Co-Owners: @andrewnguyen22

### PR DESCRIPTION
**Pull request by h5law**
## Description

This PR addresses issue #102. It consolidates `Account` and `Pool` to use a shared interface `ProtocolAccountSchema` that generalises the functions that both actors use.

The `ProtocolAccountSchema` interface is implemented by the `baseProtocolAccountSchema` struct - following the same pattern laid out by the `ProtocolActorSchema` and `baseProtocolActorSchema`. Two new types are then created - `Account` and `Pool` these are instances of the `baseProtocolAccountSchema` struct and are initiated with their relevant `SQL` table names, height constraints and column names.

The functions in the `PersistenceRWContext` for both `Account` and `Pool` utilities are now generalised to use shared logic dramatically reducing the code footprint.

In addition to this all relevant unit tests have been updated and usages of these functions are working correctly. Also regarding issue #149 `address []bytes` has been removed from `InsertPool`.

## Issue

Fixes #102 

## Type of change

Please mark the relevant option(s):

- [x] New feature, functionality or library
- [ ] Bug fix
- [x] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [ ] Other 

## List of changes

- Create `ProtocolAccountSchema` interface and `baseProtocolAccountSchema` struct that implements this interface
- Create `Account` and `Pool` types that share the common functionality of `baseProtocolAccountSchema`
- Generalise all `Account` and `Pool` functions to use shared logic
- Remove `address []byte` from `InsertPool()`
- Seperate `shared_sql.go` and `types/shared_sql.go` into `account_shared_sql.go` and `actor_shared_sql.go` as well as `types/account_shared_sql.go` and `types/actor_shared_sql.go` respectively

## Testing

- [x] `make develop_test`
- [x] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md) w/ all of the steps outlined in the `README`

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have tested my changes using the available tooling
- [x] I have updated the corresponding CHANGELOG

### If Applicable Checklist
- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

**Comment by h5law** [comment]
@Olshansk I have also addressed the discussion comment in the `persistence/account.go`  file and included the code for the relevant change if desired - commented out though. The current functionality is identical to how it was prior - only using shared logic instead.

File: persistence/shared_sql.go [hunk 1]
```diff
@@ -83,6 +86,19 @@ func (p *PostgresContext) GetActorsUpdated(actorSchema types.ProtocolActorSchema
 	return
 }
 
+func (p *PostgresContext) getAccountAmount(accountSchema types.ProtocolAccountSchema, address string, height int64) (amount string, err error) {
```
Comment (Olshansk) [inline persistence/shared_sql.go:89]: Can the functions here go into a separate `account_shared_sql`?

Might mean we have to rename this to `actor_shared_sql`

File: persistence/genesis.go [hunk 2]
```diff
@@ -46,8 +46,7 @@ func (m *persistenceModule) populateGenesisState(state *genesis.GenesisState) {
 		}
 	}
 	for _, pool := range state.GetPools() {
-		poolNameBytes := []byte(pool.GetAddress())
-		err = rwContext.InsertPool(pool.GetAddress(), poolNameBytes, pool.GetAmount())
+		err = rwContext.InsertPool(pool.GetAddress(), pool.GetAmount())
```
Comment (Olshansk) [inline persistence/genesis.go:49]: Please add a comment that the `address` for pools is actually the name.

You can find more details on it in [here](https://github.com/pokt-network/pocket-network-protocol/blob/issues/23/aat/utility/README.md#373-pool) or see the `genesis.json` file for reference:

```json
{
  "accounts": [
    {
      "address": "6f66574e1f50f0ef72dff748c3f11b9e0e89d32a",
      "amount": "100000000000000"
    },
  ],
  "pools": [
    {
      "address": "DAO",
      "amount": "100000000000000"
    },
  ],
```

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: persistence/shared_sql.go [hunk 3]
```diff
@@ -83,6 +86,19 @@ func (p *PostgresContext) GetActorsUpdated(actorSchema types.ProtocolActorSchema
...
+	ctx, tx, err := p.getCtxAndTx()
+	if err != nil {
+		return
+	}
+	amount = defaultAccountAmountStr
```
[the remark this answers is not in the document]
Comment (h5law) [reply persistence/shared_sql.go:94]: @Olshansk Do you mean check the default is zero?
These two tests already exist: `TestDefaultNonExistentPoolAmount` and `TestDefaultNonExistentAccountAmount`

Or a standalone test for the `getAccountAmount` function? The other account/pool tests heavily use this function but I can add one if needed.

File: persistence/account.go [hunk 4]
```diff
@@ -28,14 +28,11 @@ func (p PostgresContext) SubtractAccountAmount(address []byte, amount string) er
...
-// DISCUSS(team): If we are okay with `GetAccountAmount` return 0 as a default, this function can leverage
-// `operationAccountAmount` with `*orig = *delta` and make everything much simpler.
 func (p PostgresContext) SetAccountAmount(address []byte, amount string) error {
-	return p.insertAccount(types.Account, hex.EncodeToString(address), amount)
-	// return p.operationAccountAmount(types.Account, hex.EncodeToString(address), amount, func(orig, delta *big.Int) error {
-	//	orig.Set(delta)
-	//	return nil
-	// })
+	return p.operationAccountAmount(types.Account, hex.EncodeToString(address), amount, func(orig, amount *big.Int) error {
+		orig.Set(amount)
```
Comment (Olshansk) [inline persistence/account.go:33]: Any chance you can add a unit test for this? One for pool and one for account?

File: persistence/types/account_shared_sql.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
+package types
+
```
Comment (Olshansk) [inline persistence/types/account_shared_sql.go:2]:
```suggestion
// CLEANUP: Move SQL specific business logic into a `sql` package under `persistence`
```

@h5law Also, feel free to resolve comments (i.e. suggestions like these) after you apply them (either through github or yourself). There's always some inherent trust between the reviewer and reviewee.

File: persistence/types/base_account.go [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+import (
+	"fmt"
+)
+
+var _ ProtocolAccountSchema = &BaseProtocolAccountSchema{}
+
+// Implements the ProtocolAccountSchema interface that can be shared across both Accounts and Pools
+// allowing for the generalisation and sharing of code between these two entities
+type BaseProtocolAccountSchema struct {
+	// SQL Tables
```
[the remark this answers is not in the document]
Comment (Olshansk) [reply persistence/types/base_account.go:12]: Bump?

File: persistence/shared_sql.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (Olshansk) [reply persistence/shared_sql.go]: Ah, I didn't know!

**Review by Olshansk — APPROVED** [review]
@h5law LGTM! 

Just a couple things regarding review:

1. Make sure to re-request review when you need me to take another look
2. Don't hesitate to resolve the comments yourself if your confident you addressed the comment
3. Just double checking that LocalNet works fine? If so, go ahead and squash-and-merge!

File: persistence/types/base_account.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (h5law) [reply persistence/types/base_account.go:12]: Missed this one done now 👍🏼

**Comment by h5law** [comment]
@Olshansk 

>     1. Make sure to re-request review when you need me to take another look
>     2. Don't hesitate to resolve the comments yourself if your confident you addressed the comment

I thought I did these but will make sure going forward 🫡
 
>     3. Just double checking that LocalNet works fine? If so, go ahead and squash-and-merge!

All good - will update CHANGELOGs and merge after that

**Pull request merged**